core/src/fetch_stage.rs

```diff
@@ -234,6 +240,13 @@ impl FetchStage {
                 tpu_vote_stats.report();
                 tpu_forward_stats.report();
 
+                if let Some(collector) = &prometheus_collector {
+                    let stats = {
+                        tpu_stats.total_stats.lock().unwrap().clone()
+                    };
+                    collector.lock().unwrap().save_tpu_receiver_stats(stats)
```

Comment:
Hmm, I see, so that would basically result in changing how the logic of current Solana metrics are implemented. I would be a bit hesitant to change this Solana metrics logic because to make it consistent we would likely need to do it in all places this pattern is followed (which seems to be a lot).

> Not sure if this is feasible in practice though, it looks like there is something going on with accumulation over multiple threads?

Yes, that is what is happening.

> Also, having the separate fields as atomic ints can lead to surprising results. For example, if you are trying to compute an error rate, and you have counters num_requests and num_errors

Yeah there are definitely possible inconsistencies, but I guess that is really common in metrics. Using the example of `num_requests` and `num_errors` in some other application, `num_requests` is likely to be incremented at the beginning of handling the request while `num_errors` for obvious reasons can only be incremented later, so it won't be atomic operation.

P.S. I now want to see this whole application for teleporting goats 🤣

Reply:
> num_requests is likely to be incremented at the beginning of handling the request while num_errors for obvious reasons can only be incremented later, so it won't be atomic operation.

That by itself is not an issue, since you still couldn’t get a >100% error rate. If you increment the error counter before the request counter, then you could get a >100% error rate, which is really counter-intuitive. (Or if you read the request counter before the error counter.)

> I now want to see this whole application for teleporting goats

You might enjoy these :-)
https://crbug.com/31482
https://codereview.chromium.org/543045